### PR TITLE
Use hidden-args in Morganey

### LINKED ISSUE #233: Use hidden-args
Use [hidden-args](https://github.com/keddelzz/hidden-args) to remove boilerplate of functions with an nested tail-recursive helper-function. To be able to use [hidden-args](https://github.com/keddelzz/hidden-args), two new dependencies have to be added to Morganey:
- [hidden-args](https://github.com/keddelzz/hidden-args) itself
- The compiler plugin [macroparadise](https://github.com/scalamacros/paradise)

### PR DESCRIPTION
**Pull request by keddelzz**
closes #233

### CODE REVIEW

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 1]
```diff
@@ -78,15 +79,14 @@ case class LambdaFunc(parameter: LambdaVar, body: LambdaTerm) extends LambdaTerm
...
-    def go(func: LambdaFunc, acc: List[LambdaVar]): (List[LambdaVar], LambdaTerm) =
-      func match {
-        case LambdaFunc(v, f: LambdaFunc) => go(f, v :: acc)
-        case LambdaFunc(v, b)             => ((v :: acc).reverse, b)
-      }
-    go(this, Nil)
-  }
+  @hiddenargs
+  @tailrec
+  private[ast] def nestedFunctions(@hidden func: LambdaFunc     = this,
```
Comment (rexim) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala]: I'm sorry but this is just straight up ugly. Why not get rid of that redundant argument and pattern match the body of the LambdaFunc?

``` scala
def nestedFunctions(@hidden acc: List[LambdaVar] = Nil): (List[LambdaVar], LambdaTerm) =
  body match {
    case f: LambdaFunc => f.nestedFunctions(parameter :: acc)
    case _             => ((parameter :: acc).reverse, body)
  }
```

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 2]
```diff
@@ -105,15 +105,14 @@ case class LambdaApp(leftTerm: LambdaTerm, rightTerm: LambdaTerm) extends Lambda
...
-    def go(app: LambdaApp, acc: List[LambdaTerm]): (LambdaTerm, List[LambdaTerm]) =
-      app match {
-        case LambdaApp(l: LambdaApp, r) => go(l, r :: acc)
-        case LambdaApp(l, r)            => (l, r :: acc)
-      }
-    go(this, Nil)
-  }
+  @hiddenargs
+  @tailrec
+  private[ast] def nestedApplications(@hidden app: LambdaApp        = this,
```
Comment (rexim) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala]: Same as for nestedFunctions

**Review by rexim — CHANGES_REQUESTED** [review]

**Comment by rexim** [comment]
@keddelzz, thank you! Looks much better! Sorry if I was too rude. I wanna perform some manual tests as well, but I'm already falling asleep. Gonna approve your PR tomorrow.

**Comment by keddelzz** [comment]
> Sorry if I was too rude.

No need to apologize! You were absolutely right. It was ugly. :laughing:

**Review by ForNeVeR — APPROVED** [review]

**Review by rexim — APPROVED** [review]

**Comment by rexim** [comment]
@keddelzz thanks for the changes!
@ForNeVeR thanks for the review!

**Pull request merged**